### DISCUSSION TITLE
Register CommonModel to RootModel got Ts error: circularly references itself in mapped type 'ExtractRematchDispatchersFromModels '.

### DISCUSSION THREAD

**Discussion by 6daya** [rematch/rematch#838] [Q&A]
First, i created my `commonModel` which looks like:

```ts
// common model
const myCommonModel = createModel<CommonModel>()({
    state: {},
    reducers: {},
    effects: dispatch => ({
    }),
});

export interface CommonModel extends Models<CommonModel> {
    myCommon: typeof commonModel;
}
```

and here’s my actual `pageModel`, which wanna register the `commonModel`.

```ts
export interface RootModel extends Models<RootModel> {
    myCommon: typeof myCommonModel;
}
```

And this will get a ts error: `TS2615: Type of property 'myCommon' circularly references itself in mapped type 'ExtractRematchDispatchersFromModels '.`

Or should it be registered on `models` which is inside `init` function options.

```ts
export const store = init<RootModel>({
    models: {
        common: myCommonModel,
    },
});

```

but it also got a ts error which confused me:

`
TS2322: Type '{ name?: string | undefined; state: {}; reducers: {}; baseReducer: Reducer<{}, AnyAction>; effects: (dispatch: RematchDispatch<CommonModel>) => {}; }' is not assignable to type 'Model<RootModel, any, any>'.   Types of property 'effects' are incompatible.     Type '(dispatch: RematchDispatch<CommonModel>) => {}' is not assignable to type 'ModelEffects<RootModel> | ModelEffectsCreator<RootModel> | undefined'.       Type '(dispatch: RematchDispatch<CommonModel>) => {}' is not assignable to type 'ModelEffectsCreator<RootModel>'.         Types of parameters 'dispatch' and 'dispatch' are incompatible.           Type 'RematchDispatch<RootModel>' is not assignable to type 'RematchDispatch<CommonModel>'.             Property 'myCommon' is missing in type 'Dispatch<AnyAction> & ExtractRematchDispatchersFromModels<RootModel>' but required in type 'ExtractRematchDispatchersFromModels<CommonModel>'.
`

This part of the usage is incomplete in the doc.

how can i deal with this? thanks.

**Comment by semoal** [comment]
Could you provide a little repo with a reproduction? In that way i'm 100% percent that we could help you 👍🏻

**Comment by 6daya** [comment]
thanks a lot.

and heres my repo demo: https://stackblitz.com/edit/react-ts-x2k2sb?file=models/index.ts

main logic is under `model`folder, `store.ts` and `test.ts`

**Comment by tianzhich** [comment]
In your `commonModel `, put the `RootModel` but not `CommonModel` as the generic of `createModel`: 

```ts
// common model
const myCommonModel = createModel<RootModel>()({
    state: {},
    reducers: {},
    effects: dispatch => ({
    }),
});

// You could remove this cause unused.
export interface CommonModel extends Models<CommonModel> {
    myCommon: typeof commonModel;
}
```

checkout: https://stackblitz.com/edit/react-ts-wh31su?file=models%2Fcommon.ts

**Comment by 6daya** [comment]
the `commonModel`  and the `RootModel` are not related in any way, 
thus it will not use the `RootModel` as the generic of createModel

`commonModel` will not know which model register it。

or maybe use Generics function to pass the `RootModel`? i am not sure if its the best way

**Comment by tianzhich** [comment]
The relation is that your `commonModel` need to know the `RootModel` because you could dispatch actions of all of your models from here: 

```ts
effects: dispatch => ({
}),
```

I also found online editors seem not to use the newest version of typescript, you can see https://github.com/stackblitz/core/issues/483 which is your repro editor, and https://github.com/codesandbox/codesandbox-client/issues/942 which is [codesandbox](https://codesandbox.io/).

So maybe it was an issue with the old typescript. I created some code snippets in [TS Playground](https://www.typescriptlang.org/play?ts=4.1.2#code/KYDwDg9gTgLgBASwHY2FAZgQwMbDgWQgBNgAbAZwB4AVQkiuUVJI8g4sq2jigPjgC8cAErBs0IpXIwoyAOYAaOJiQBPXvwDeAWABQcOAG0A1sFUAuONNlI5AXUt0yNJ3z0BfPXtCRYiFGhYuOz0AKLo6GIwVHoG3PRsTMAsbK5cafxCouJQktbySirqsXBaJSZmlvm2DiFk4ZHYMC485LweXro+0PAk2KSYUHgwqmB4ogC2mDDYABYAIgjkYNNzlCXxnIwgzKx1FC0JmSJiElIyBcpq7fqlgnChO1A4MJOrC0srM7No5ABiUAgEzSh04N284B6cD6AyGcBGYweTxeb2+i2W71+AKBII2aW2u1SrVBfHu2TO1UUV2Kt2OOluBkMEx4AGkzP44KZVBB0HBNhRavzyEzWWY7ATkntXCS2iUDAB+JEyFHAKZoz6YqD-QETBpRdKtEX0NmqOyGADkwAi+vNdiUQpuBgMliQwAAbmgSp5dBDfL0xLDhqM8I9lU1UXN0V85lidXqmuQAPIAIwAVlF1rdqPHohKUvscwajgo8a081LiQ6yadcucbFSio67kJ6U6jFbGjATRyuTy+YXaqHnuHVe8o5rsbrrU1M224oXDB2oia7XKnUK103vb6oTDBkHEUOVWrIxrvmhJznZ-3p-AkvnXFeHSWs-j7xWEjLjuTa5TCtcSmObNb3LNgAAoAEpBH4ZBIigPlRBgABXKAkBKRUcwjD4MXPKAaEQlCkCbSxgM7UC4DAlZVFICBMCISxYLQPkAAVMGo2iiCggQYKQOCEOAZDUPQh5byw8dcPwgTCPtVj2Lo4ibzI99wKomi6IY3imOoWS1KIJRpGmYBLCKLieL46gCKE24MNE0d1RwmM8IsqTUJktjdIU10PSgTpuj8Pc4QRENbJPbDox+JzLKQe4ijcuSiHuN0IAQIhjkMDYdI4vRxWUowkpSuxhLAyDoP4wSkCggAyOBWydJYc0sGQkOAL0SksYrVI4kjMro0yysIqqarXerb0aqBmtan0ukhfyA33eFgyVYdXjs08HIi7UgULa9SP1cjpQdC0lwTW0XziN8dklIlPyrLIazyC5bH-GkDCAwtyLAgA6b7BjkchjLUQw7D6xj4OEYThH2nhtodYSjxHULxMczap07JM0wzYR7QyNq4C8z1bhI97csfW8izBOHkQRscz2Ry8yZTdMZ12hNsdaBT8qIXyZv9fp5qCgsyYAYSGaZoB2i7CX2cnSTunIHvrZ6m2OMCiFp2ZLDE9Wvz60m0a-ToBdcEXgDF+ChEoYR8Chz8rY0EruLga9Qht+oyct-B+AAH0FtGTbNj3eDOuAAGVg94MDmUsWqDCQTAJmAeUqkeuQ1wM1BLBDtdjuiJOHi9PqY7x+PE+T+s05gQzM+zsnLFCSbub9OBxCQaRm9F1BXEcHh-ZgaB7gd-hI4gCCAdUUrapbtvNEYMn7k0dw4EXoRmU6AwhnKoaGTgb7PuZYODBz8hg+9bdpqbgKD3GVbZmECAIBgEPK9QCWyxJysMmreW60uRtg+zZEksrrSy-PPb0r17jwxWqFJ+hlJwgiFPaI8ONz67jmoFRaUCsKwNQPA4kpYEiuwOLdE438-zUiDhsZBb9LoPg-uzMBgF55riNGQbsyBORmD7EKQUaRWGkBXBadOwBbRejgJVEoRd+HsOir2XkADlRpEFNQhI0ixRCOfiIwqtwz5+V5oGBaiI9YZgIVsd+N1P5ywpCnJWTChCUXcl1akShAQPxwUZE4oU75uM0TrUqRRG5QmQKgDAOA8DSlMQwcxnBQFWN-DYih-93ExTUEw2qccE550pCUYRJF3ElCPnnYxCZQE+2KdEXu4tYa6F0TzZuEBW7wAmKoIWQJmTRSENgDuwBpStImO01wEcIJgVqrkmq7hg5H0sGrdapURk107PMVUEBIJFwMFPCApBgCfRonIMClpbxLOZOaCCa4JkFw8BBAA3IEvwwTAhhLgH0gZPAiFUGeQ0wZW8DDNI+UgRqwY+y-LaQ0m5NTbnwHuaE4I3iYCuDeZbe+cKeBlFuAAejRXACKeApiqGTDi8eD8IpwBWHIZZCQSglAxYgOAAB3FQcc4BDDkEsEJ9T+kNLgMyegWK0B4A4TAWYSwmVIq5TwEowKOX-MMcAIFLSQVIDBbUpuU8mmtE1ki+FLYJXyqlS+dwVygA) in which your problem not occur.

**Comment by tianzhich** [comment]
Have you encountered this on your local or just online editors? You can set your local editor to use the newest typescript and test it again.

**Comment by 6daya** [comment]
it occurred on both my local and online editors, and  thanks for reminding the version, i have tried `^4.1.2` and `^3.8.3`, but both of them show the same error, which is `TS2615: Type of property 'myCommon' circularly references itself in mapped type 'ExtractRematchDispatchersFromModels '.`

**Comment by 6daya** [comment]
The relation is that your commonModel need to know the RootModel because you could dispatch actions of all of your models from here
------
@tianzhich 

maybe it's not, cause the commonModel will not use any states or effects of my pageModel. pageModel is just register the comonModel on itself, and use it as a common ability.

for example, 

in a multi-page application, every page has its own store and model.And more then one page need a rank-list, they all need the `rankModel`. Thus we extract this part of the logic from the page level to the common level.

and this why we need a common model, and this common model will not know the which page model use it, and will not know which `RootModel` it should use as the generic of createModel

**Comment by 6daya** [comment]
@tianzhich bro, iI was also worked in  Kwai before. l59733984, my wechat

**Comment by 6daya** [comment]
the code in TS Playground works fine in my local.but the origin code still hint the same error

**Comment by tianzhich** [comment]
Of course, you can use your `commonModel` as a common model and don't need to dispatch actions of other models from here.

But rematch collects all your actions and register them on `dispatch` param. So we add this typing support.

It's nice to meet former colleagues on Github. I'll connect you on WeChat soon.

**Comment by semoal** [comment]
For me has sense that typescript is failing because you're passing wrong the generics. And if your commonmodel doesn't need to know another models, just don't type it.
